kolibri/plugins/hooks.py

```diff
@@ -1,59 +1,327 @@
+        If you wonder whether you need them, you don't (the people who actually
+        need them know with certainty that they need them, and don't need an
+        explanation about why).
+
+        - Tim Peters
+
     """
-    return REGISTERED_HOOKS[hook]
+
+    def __new__(cls, name, bases, attrs):
+        """
+        Inspired by Django's ModelBase, we create a dynamic type for each hook
+        definition type class and add it to the global registry of hooks.
+        """
+
+        super_new = super(KolibriHookMeta, cls).__new__
+
+        # Also ensure initialization is only performed for subclasses of Model
+        # (excluding Model class itself).
+        parents = [b for b in bases if isinstance(b, KolibriHookMeta)]
+
+        # If there isn't any parents, it's the main class of everything
+        # ...and we just set some empty options
+        if not parents:
+            base_class = super_new(cls, name, bases, attrs)
+            base_class.add_to_class('_meta', Options(None))
+            base_class.add_to_class('_parents', [])
+            return base_class
+
+        # Create the class.
+        module = attrs.pop('__module__')
+        new_class = super_new(cls, name, bases, {'__module__': module})
+
+        attr_meta = attrs.pop('Meta', None)
+        abstract = getattr(attr_meta, 'abstract', False)
+        if not attr_meta:
+            meta = getattr(new_class, 'Meta', None)
+        else:
+            meta = attr_meta
+
+        # Meta of the base object can be retrieved by looking at the currently
+        # set _meta object... but we don't use it...
+        # base_meta = getattr(new_class, '_meta', None)
+
+        # Add all attributes to the class.
+        for obj_name, obj in attrs.items():
+            new_class.add_to_class(obj_name, obj)
+
+        new_class.add_to_class('_meta', Options(meta))
+        new_class.add_to_class('_parents', parents)
+
+        if not abstract:
+            logger.debug("Registered hook class {}".format(new_class))
+            for parent in new_class._parents:
+                parent.register_hook(new_class)
+
+        return new_class
+
+    def add_to_class(cls, name, value):
+        setattr(cls, name, value)
+
+    def register_hook(cls, child_hook):
+        cls._meta.registered_hooks.add(child_hook)
+        for parent in cls._parents:
+            parent.register_hook(child_hook)
+
+
+class KolibriHook(six.with_metaclass(KolibriHookMeta)):
+    """
+    WIP!!!
+
+    To use it, extend it. All hooks in kolibri extends from this.
+    Example is in the main description of ``kolibri.plugins.hooks``.
+    """
+
+    def __init__(self):
+        # We have been initialized. A noop. But inheriting hooks might want to
+        # pay attention when they are initialized, since it's an event in itself
+        # signaling that whatever that hook was intended for is now being
+        # yielded or rendered.
+        pass
```

Comment:
Can you give an example of when we would actually want to override `__init__` for a hook class?